fdb-record-layer-lucene/src/main/java/com/apple/foundationdb/record/lucene/LuceneSpellcheckRecordCursor.java

```diff
@@ -0,0 +1,234 @@
+import org.apache.lucene.index.DirectoryReader;
+import org.apache.lucene.index.IndexReader;
+import org.apache.lucene.index.Term;
+import org.apache.lucene.search.spell.DirectSpellChecker;
+import org.apache.lucene.search.spell.SuggestWord;
+//import org.slf4j.Logger;
+//import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Executor;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+import static com.apple.foundationdb.record.lucene.DirectoryCommitCheckAsync.getOrCreateDirectoryCommitCheckAsync;
+import static com.apple.foundationdb.record.lucene.IndexWriterCommitCheckAsync.getIndexWriterCommitCheckAsync;
+import static java.util.Comparator.comparing;
+
+public class LuceneSpellcheckRecordCursor implements BaseCursor<IndexEntry> {
+
+    // TODO: log some stuff.
+    // private static final Logger LOGGER = LoggerFactory.getLogger(LuceneSpellcheckRecordCursor.class);
+    @Nonnull
+    private final Executor executor;
+    @Nonnull
+    private final IndexMaintainerState state;
+    private final int limit;
+    @Nonnull
+    private final String wordToSpellCheck;
+    @Nonnull
+    private final DirectSpellChecker spellchecker;
+    @Nullable
+    private final FDBStoreTimer timer;
+
+    @Nullable
+    private List<IndexEntry> spellcheckSuggestions = null;
+    private int currentPosition = 0;
+    @Nullable
+    private final Tuple groupingKey;
+    private final List<String> fields;
+
+
+    public LuceneSpellcheckRecordCursor(@Nonnull final String value,
+                                        @Nonnull final Executor executor,
+                                        final ScanProperties scanProperties,
+                                        @Nonnull final IndexMaintainerState state,
+                                        @Nullable Tuple groupingKey,
+                                        final String[] fieldNames) {
+        if (value.contains(":")) {
+            String[] fieldAndWord = value.split(":", 2);
+            // only check the validity of the field if the documents are not grouped.
+            if (groupingKey != null
+                    && groupingKey.size() == 0
+                    && Arrays.stream(fieldNames).noneMatch(name -> name.equals(fieldAndWord[0]))) {
+                throw new RecordCoreException("Invalid field name in Lucene index query")
+                        .addLogInfo(LogMessageKeys.FIELD_NAME, fieldAndWord[0])
+                        .addLogInfo(LogMessageKeys.INDEX_FIELDS, fieldNames);
+            }
+            fields = List.of(fieldAndWord[0]);
+            wordToSpellCheck = fieldAndWord[1];
+        } else {
+            // TODO in a grouped document this doesn't work because the field names all include the grouping key
```

Comment:
Do they? Doesn't the grouping key select a whole different Lucene directory?

Reply:
I am not sure I understand this comment. Does the grouping key necessarily be part of the Lucene field name?